packages/SwingSet/tools/run-utils.ts

```diff
@@ -0,0 +1,152 @@
+      case 'unresolved':
+        throw Error(`unresolved value for ${kpid}`);
+      default:
+        throw Fail`unknown status ${status}`;
+    }
+  };
+
+  type EVProxy = typeof E & {
+    sendOnly: (presence: unknown) => Record<string, (...args: any) => void>;
+    vat: (name: string) => Record<string, (...args: any) => Promise<any>>;
+  };
+
+  // IMPORTANT WARNING TO USERS OF `EV`
+  //
+  // `EV` presents an abstraction that can be used (within tests only!) to get
+  // much of the convenience with respect to messaging that `E` provides in
+  // normal code.  However, this convenience comes with a huge caveat that all
+  // users of this convenience feature MUST keep in mind.
+  //
+  // A test can drop messages onto the kernel's run queue using the
+  // `controller.queueToVatRoot` and `controller.queueToVatObject` methods.
+  // These are synchronous operations which merely place messages onto the run
+  // queue without causing execution.  Execution, on the other hand, is
+  // initiated by calling `controller.run`, which will cause the kernel to begin
+  // delivering messages to vats from the run queue, continuing until the run
+  // queue is exhausted.  HOWEVER, exhaustion of the run queue, which resolves
+  // the result promise returned by the `run` call, IS NOT coupled in any causal
+  // way to the resolution of result promises associated with the individual
+  // queued messages themselves.  The status and resolution values of these
+  // promises can be synchronously queried (by kpid) via the
+  // `controller.kpStatus` and `controller.kpResolution` methods once `run` has
+  // completed.  These queries are only available once the swingset has
+  // reqlinquished agency, i.e., when the work initiated by `controller.run` has
+  // finished.  At that point, nothing is going on inside the kernel, and
+  // nothing WILL be going on inside the kernel until a subsequent call to
+  // `controller.run`, which in turn will only have an effect if additional
+  // messages have been placed on the kernel run queue in the meantime.  You MAY
+  // NOT call `queueToVatRoot`, `queueToVatObject`, `kpStatus`, or
+  // `kpResolution` while run queue execution, triggered by a call to `run`, is
+  // in progress
+  //
+  // The functionality made available by `EV` looks similar to that provided by
+  // `E`, but it is very much not.  When you send a message using `EV`, it
+  // places the message onto the kernal run queue and then immediately invokes
+  // `controller.run`.  When the result of `run` resolves, the kpid returned by
+  // the message enqueueing operation is queried.  If at that time the promise
+  // it identifies is resolved (or rejected), the value it was resolved (or
+  // rejected) to is used as the result from the `EV` invocation.  However, if
+  // it is still pending at that time, `EV` will throw an exception and your
+  // test will abort.  This means that if you initiate some operation via an
```

Comment:
I don't think this is true (at least the first part), since `queueAndRun` is async and therefore transforms errors into rejections.

Reply:
Good point, though I'm not sure how much it matters (nevertheless I'll clean up the language to clarify).  In most cases, Ava will turn the rejection back into an error, but if your test was already expecting too see a rejection it will indeed see a rejection, but not the one it was expecting; that will consequently be reflected as a regular test failure rather than an abort.  You'll still get a diagnostic that something is wrong, but it could be confusing if you don't know about this failure mode.